src/ripple/app/tx/impl/URIToken.cpp

```diff
@@ -0,0 +1,819 @@
+                                      "valid utf-8 string.";
+                return temMALFORMED;
+            }
+
+            break;
+        }
+
+        case ttURITOKEN_CANCEL_SELL_OFFER:
+        case ttURITOKEN_BURN: {
+            if (flags & tfURITokenNonMintMask)
+                return temINVALID_FLAG;
+            break;
+        }
+
+        case ttURITOKEN_BUY:
+        case ttURITOKEN_CREATE_SELL_OFFER: {
+            if (flags & tfURITokenNonMintMask)
+                return temINVALID_FLAG;
+
+            auto amt = ctx.tx.getFieldAmount(sfAmount);
+
+            if (!isLegalNet(amt) || amt.signum() < 0)
+            {
+                JLOG(ctx.j.warn()) << "Malformed transaction. Negative or "
+                                      "invalid amount/currency specified.";
+                return temBAD_AMOUNT;
+            }
+
+            if (badCurrency() == amt.getCurrency())
+            {
+                JLOG(ctx.j.warn()) << "Malformed transaction. Bad currency.";
+                return temBAD_CURRENCY;
+            }
+
+            if (tt == ttURITOKEN_BUY)
+                break;
+
+            if (amt == beast::zero && !ctx.tx.isFieldPresent(sfDestination))
+            {
+                JLOG(ctx.j.warn()) << "Malformed transaction. "
+                                   << "If no sell-to destination is specified "
+                                      "then a non-zero price must be set.";
+                return temMALFORMED;
+            }
+            break;
+        }
+
+        default:
+            return tefINTERNAL;
+    }
+
+    return preflight2(ctx);
+}
+
+TER
+URIToken::preclaim(PreclaimContext const& ctx)
+{
+    std::shared_ptr<SLE const> sleU;
+    uint32_t leFlags;
+    std::optional<AccountID> issuer;
+    std::optional<AccountID> owner;
+    std::optional<STAmount> saleAmount;
+    std::optional<AccountID> dest;
+    std::shared_ptr<SLE const> sleOwner;
+
+    if (ctx.tx.isFieldPresent(sfURITokenID))
+    {
+        sleU = ctx.view.read(
+            Keylet{ltURI_TOKEN, ctx.tx.getFieldH256(sfURITokenID)});
+        if (!sleU)
+            return tecNO_ENTRY;
+
+        leFlags = sleU ? sleU->getFieldU32(sfFlags) : 0;
+        owner = sleU->getAccountID(sfOwner);
+        issuer = sleU->getAccountID(sfIssuer);
+        if (sleU->isFieldPresent(sfAmount))
+            saleAmount = sleU->getFieldAmount(sfAmount);
+
+        if (sleU->isFieldPresent(sfDestination))
+            dest = sleU->getAccountID(sfDestination);
+
+        sleOwner = ctx.view.read(keylet::account(*owner));
+        if (!sleOwner)
+        {
+            JLOG(ctx.j.warn()) << "Malformed transaction: owner of URIToken is "
+                                  "not in the ledger.";
+            return tecNO_ENTRY;
+        }
+    }
+
+    AccountID const acc = ctx.tx.getAccountID(sfAccount);
+    uint16_t tt = ctx.tx.getFieldU16(sfTransactionType);
+
+    switch (tt)
+    {
+        case ttURITOKEN_MINT: {
```

Comment:
what's the rationale behind using a switch here (and above)? avoiding a new call stack frame?

Reply:
a switch is the correct structure to use here:
* preflight checks are identical between two pairs of the five txn types, the switch allows fall through logic.
* the code is very clear, using another structure would make it less clear.